contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_binary, to_binary, Binary, Deps, DepsMut, Empty, Env, IbcMsg, MessageInfo, Response,
+    StdResult, SubMsg, WasmMsg,
+};
+use cw2::set_contract_version;
+
+use crate::error::ContractError;
+use crate::helpers::{
+    burn, get_class, get_nft, get_owner, has_class, transfer, INSTANTIATE_CW721_REPLY_ID,
+};
+use crate::ibc::NonFungibleTokenPacketData;
+use crate::msg::{ExecuteMsg, IbcAwayMsg, InstantiateMsg, QueryMsg};
+use crate::state::{
+    UniversalNftInfoResponse, CHANNELS, CLASS_ID_TO_CLASS_URI, CLASS_ID_TO_NFT_CONTRACT,
+    CW721_ICS_CODE_ID, ESCROW_CODE_ID, NFT_CONTRACT_TO_CLASS_ID,
+};
+
+const CONTRACT_NAME: &str = "crates.io:cw-ics721-bridge";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    CW721_ICS_CODE_ID.save(deps.storage, &msg.cw721_ics_code_id)?;
+    ESCROW_CODE_ID.save(deps.storage, &msg.escrow_code_id)?;
+
+    Ok(Response::default()
+        .add_attribute("method", "instantiate")
+        .add_attribute("cw721_code_id", msg.cw721_ics_code_id.to_string())
+        .add_attribute("escrow_code_id", msg.escrow_code_id.to_string()))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Transfer {
```

Comment:
Per convos about architecture, these should only be called after you are in ibc_receive_packet right? Since two main entry points for should be 1) a transfer nft message sent when user does a transfer for an nft from one contract to ics721 ownership and 2) the ibc packet receive method?

Reply:
This deals with providing a ICS 721 interface for cw721 NFTs. This test might help clear it up: https://github.com/public-awesome/ics721/blob/c1d19347ba00aebee72f5c64b6a9fe5a066b93ac/contracts/cw-ics721-bridge/src/integration_tests.rs#L143-L154

Reply:
We implement the "NFT asset manager" interface from the specification as part of the bridge contract - hence the `Transfer` and `Burn` here.

Reply:
👍 makes sense